Fix #712, allow overwriting the public IP of a Kubernetes node

alvaroaleman opened the pull request:
## Description

This PR fixes #712 by allowing to overwrite the public ip of  a Kubernetes node via the `flannel.alpha.coreos.com/public-ip-overwrite` annotation.

Can anyone point me to the appropriate place for docs, is it just `Documentation/Configuration.md`?
Right now the other annotations aren't documented in this repo.

## Todos
- [x] Tests
- [x] Documentation

tomdee commented:
@alvaroaleman Thanks for this change. If you can add some documentation and ideally a test, I can get it merged. For the documentation, I would suggest adding an "Annotations" section to https://github.com/coreos/flannel/blob/master/Documentation/kubernetes.md

alvaroaleman commented:
Added docs. Writing a test seems to be rather complicated, because there are none for the kube subnet manager yet.

tomdee commented:
I added some tests for kube subnet manager recently - https://github.com/coreos/flannel/blob/master/dist/functional-test-k8s.sh

But I agree, they are still quite complicated!

alvaroaleman commented:
Thanks for the hint @tomdee, I added a test.

I also moved the check for `ls /run/flannel/subnet.env` into the `start_flannel` function, since it is a generic test if flannel startup was finished and not specific to the other ping tests in the suite.

alvaroaleman commented:
Recreated the PR because travis was failing after having issues pulling the etcd image.

The e2e tests run twice btw, because they are called in the travis file and from within the `test` maketarget

tomdee commented:
OK, this looks good. It would be great if you could add a little more information to the documentation, such as _why_ a user might want to set this, and any other annotations that can be set. 

Before I merge can you squash your commits so I'm just merging a single commit?

alvaroaleman commented:
Updated the explanation and squashed my commits.

As far as I can tell, the other annotations are not supposed to be altered by users thus it doesn't make sense to document them or am I wrong there?

tomdee commented:
Yes, I think you're right! The tests are now failing, I think you lost a commit when doing the squash? Hopefully you can get it back easily then I can merge.

alvaroaleman commented:
Added the missing commit back.
